src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala

```diff
@@ -0,0 +1,193 @@
+
+    blocks.get(hash) match {
+
+      case Some(_) =>
+        log.debug(s"Block (${blockId(block)}) already in queue. ")
+        None
+
+      case None if isNumberOutOfRange(number, bestBlockNumber) =>
+        log.debug(s"Block (${blockId(block)} is outside accepted range. Current best block number is: $bestBlockNumber")
+        None
+
+      case None =>
+        val parentTd = blockchain.getTotalDifficultyByHash(parentHash)
+
+        parentTd match {
+
+          case Some(_) =>
+            addBlock(block, parentTd)
+            log.debug(s"Enqueued new block (${blockId(block)}) with parent on the main chain")
+            updateTotalDifficulties(hash)
+
+          case None =>
+            addBlock(block, parentTd)
+            findClosestChainedAncestor(block) match {
+              case Some(ancestor) =>
+                log.debug(s"Enqueued new block (${blockId(block)}) to a rooted sidechain")
+                updateTotalDifficulties(ancestor)
+
+              case None =>
+                log.debug(s"Enqueued new block (${blockId(block)}) with unknown relation to the main chain")
+                None
+            }
+        }
+    }
+  }
+
+  def isQueued(hash: ByteString): Boolean =
+    blocks.get(hash).isDefined
+
+
+  /**
+    * Removes a branch going from descendant block upwards to the oldest ancestor. Shared part of branch is not removed
+    * @param descendant the youngest block to be removed
+    * @return full branch from oldest ancestor to descendant, even if not all of it is removed
+    */
+  def removeBranch(descendant: ByteString): List[Block] = {
+
+    def recur(hash: ByteString, childShared: Boolean): List[Block] = {
+      blocks.get(hash) match {
+        case Some(QueuedBlock(block, _)) =>
+          import block.header.parentHash
+
+          val isShared = childShared || parentToChildren.get(hash).exists(_.nonEmpty)
+          if (!isShared) {
+            val siblings = parentToChildren.get(parentHash)
+            siblings.foreach(sbls => parentToChildren += parentHash -> (sbls - hash))
+            blocks -= hash
+          }
+
+          block :: recur(parentHash, isShared)
+
+        case None =>
+          Nil
+      }
+    }
+
+    recur(descendant, false).reverse
+  }
+
+  /**
+    * Removes a whole subtree begining with the ancestor. To be used when ancestor fails to execute
+    * @param ancestor hash of the ancestor block
+    */
+  def removeSubtree(ancestor: ByteString): Unit =
+    blocks.get(ancestor).foreach { case QueuedBlock(block, _) =>
+      val children = parentToChildren.getOrElse(ancestor, Set.empty)
+      children.foreach(removeSubtree)
+      blocks -= block.header.hash
+      parentToChildren -= block.header.hash
+    }
+
+  /**
+    * Removes stale blocks - too old or too young in relation the current best block number
+    * @param bestBlockNumber - best block number of the main chain
+    */
+  private def cleanUp(bestBlockNumber: BigInt): Unit = {
```

Comment:
Shouldn't we remove the whole subtree when removing a block because it's old?

Reply:
Well, we could but there is tiny chance that after asking a peer for headers those "abandoned" children may get usable again. Other than that, they will eventually be removed anyway, and it's just slightly easier to remove them one by one.

I can change it to `removeSubtree` though, if you're convinced we should be doing it that way.

Reply:
My question was in case I was missing something, ie: keeping the children for a little bit longer had a specific use case. I think it's ok leaving as it (a small comment wont hurt tho :wink:)